receiver/vcenterreceiver/client.go

```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
 		resultsByRef: resultsByRef,
 	}, nil
 }
+
+// VSANQueryResults contains all returned vSAN metric related data
+type VSANQueryResults struct {
+	// Contains vSAN metric data keyed by UUID string
+	MetricResultsByUUID map[string]*VSANMetricResults
+}
+
+// VSANMetricResults contains vSAN metric related data for a single resource
+type VSANMetricResults struct {
+	// Contains UUID info for related resource
+	UUID string
+	// Contains returned metric value info for all metrics
+	MetricDetails []*VSANMetricDetails
+}
+
+// VSANMetricDetails contains vSAN metric data for a single metric
+type VSANMetricDetails struct {
+	// Contains the metric label
+	MetricLabel string
+	// Contains the metric interval in seconds
+	Interval int32
+	// Contains timestamps for all metric values
+	Timestamps []*time.Time
+	// Contains all values for vSAN metric label
+	Values []int64
+}
+
+// VSANQueryType represents the type of VSAN query
+type VSANQueryType string
+
+const (
+	VSANQueryTypeHosts           VSANQueryType = "host-domclient:*"
+	VSANQueryTypeVirtualMachines VSANQueryType = "virtual-machine:*"
+)
+
+// getLabelsForQueryType returns the appropriate labels for each query type
+func (vc *vcenterClient) getLabelsForQueryType(queryType VSANQueryType) []string {
+	switch queryType {
+	case VSANQueryTypeHosts:
+		return []string{
+			"iopsRead", "iopsWrite", "throughputRead", "throughputWrite",
+			"latencyRead", "latencyWrite", "congestion", "clientCacheHitRate",
+		}
+	case VSANQueryTypeVirtualMachines:
+		return []string{
+			"iopsRead", "iopsWrite", "throughputRead", "throughputWrite",
+			"latencyRead", "latencyWrite",
+		}
+	default:
+		return []string{}
+	}
+}
+
+// VSANQuery performs a VSAN query for the specified type across multiple clusters
+func (vc *vcenterClient) VSANQuery(
+	ctx context.Context,
+	queryType VSANQueryType,
+	clusterRefs []*vt.ManagedObjectReference,
+) (*VSANQueryResults, error) {
+	allResults := VSANQueryResults{
+		MetricResultsByUUID: map[string]*VSANMetricResults{},
+	}
+
+	if vc.vsanDriver == nil {
+		return &allResults, nil
+	}
+
+	for _, clusterRef := range clusterRefs {
+		results, err := vc.vsanQueryByCluster(ctx, queryType, clusterRef)
+		if err != nil {
+			vsanErr := vc.handleVSANError(err, queryType)
+			if vsanErr != nil {
+				return nil, vsanErr
+			}
+			continue
+		}
+
+		maps.Copy(allResults.MetricResultsByUUID, results.MetricResultsByUUID)
+	}
+
+	return &allResults, nil
+}
+
+// vsanQueryByCluster performs a VSAN query for a specific cluster
+func (vc *vcenterClient) vsanQueryByCluster(
```

Comment:
I can guess why you took the vSANDriver nil check out of this function, but this function should be able to stand on its own. So I would put it back here.

Reply:
Again, in my other PRs this function doesn't even exist...